tls/s2n_cipher_suites.c

```diff
@@ -1079,10 +1079,19 @@ struct s2n_cipher_suite *s2n_cipher_suite_from_wire(const uint8_t cipher_suite[S
 int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_CIPHER_SUITE_LEN])
 {
     notnull_check(conn);
+    notnull_check(conn->secure.cipher_suite);
+    struct s2n_cipher_suite *cipher_suite;
 
     /* See if the cipher is one we support */
-    conn->secure.cipher_suite = s2n_cipher_suite_from_wire(wire);
-    S2N_ERROR_IF(conn->secure.cipher_suite == NULL, S2N_ERR_CIPHER_NOT_SUPPORTED);
+    cipher_suite = s2n_cipher_suite_from_wire(wire);
+    S2N_ERROR_IF(cipher_suite == NULL, S2N_ERR_CIPHER_NOT_SUPPORTED);
+
+    /* Verify cipher suite sent in server hello is the same as sent in hello retry */
+    if (conn->secure.cipher_suite != &s2n_null_cipher_suite) {
```

Comment:
When would it be okay that this was the null cipher suite?

Reply:
I'm using whether or not the client cipher suite has been set as a way of differentiating between hello retry requests and server hellos. The client cipher suite should be null until parsing a hello retry request or a server hello. If it is not null, then we know that a hello retry request has been processed previously and we can check that the cipher suite being sent through the server hello has not changed. I'm not sure if there is a better way of checking that a hello retry request has been processed.

Reply:
You probably want https://github.com/awslabs/s2n/blob/master/tls/s2n_tls13.h#L29 It checks if this is a retry handshake

Reply:
Specifically what I need to check here is if this is a server hello after a hello retry request. So the check would be the combination of s2n_is_hello_retry() && s2n_hello_retry_validate(), which seemed messy and confusing to read. I can change it, I'm just not sure if it will be any easier to understand.